### PR TITLE
Add ability to select which transactions/samplers are included in report through regex

### PR DESCRIPTION
**Pull request by pmouawad**
Hello,
As per little discussion on :
https://wiki.jenkins.io/display/JENKINS/Performance+Plugin?focusedCommentId=168919627&#comment-168919627

This PR adds regexFilter which if not empty will filter which transactions are included in report.
This will enable for JMeter for example to work smoothly when using Naming Policy for default HTML report , user can use the same regex as the one configured in jmeter.reportgenerator.exporter.html.series_filter and does not need anymore to check Generate Parent Sampler in Transaction Controller to filter losing some precious information from HTML report.

For other tools, it will allow filtering transactions which is an enhancement.

I ran successfully JUnit tests
I made tests on Jenkins with Publish Performance Test report

But since bzt does not work with Jenkins on MACOSX, I was not able to test report generation after a test triggered, so you should test this part.

Thanks
Regards
Philippe M.
Note this work was sponsored by [Ubik Load Pack](https://ubikloadpack.com/)

### CODE REVIEW

**Comment by pmouawad** [comment]
Hello @undera, @artem-fedorov,

Is there an issue with PR ?

Thanks

File: docs/Changelog.md [hunk 1]
```diff
@@ -1,8 +1,11 @@
 <small>[<< Back to main page](./)</small>
 # Changelog
 
+## v3.16 (31th of January, 2019)
+- ADD: Add ability to select which transactions/samplers are included in report through regular expression. Developed by [Philippe M.](https://leanpub.com/master-jmeter-from-load-test-to-devops) and sponsored by [Ubik Load Pack](https://ubikloadpack.com)
```
Comment (artem-fedorov) [inline docs/Changelog.md]: Please, replace this link with link to your github profile

File: src/main/java/hudson/plugins/performance/reports/PerformanceReport.java [hunk 2]
```diff
@@ -176,6 +201,10 @@ public void addSample(TaurusFinalStats sample, boolean isSummaryReport) {
                             + "name properly for each http sample: skipping sample");
             return;
         }
+        if(!isIncluded(uri)) {
```
Comment (artem-fedorov) [inline src/main/java/hudson/plugins/performance/reports/PerformanceReport.java]: code style

File: src/main/java/hudson/plugins/performance/reports/PerformanceReport.java [hunk 3]
```diff
@@ -167,6 +184,14 @@ public void addSample(HttpSample pHttpSample) throws SAXException {
         totalSizeInKB += pHttpSample.getSizeInKb();
     }
 
+    private boolean isIncluded(String name) {
+        if (filterRegexPattern!=null) {
```
Comment (artem-fedorov) [inline src/main/java/hudson/plugins/performance/reports/PerformanceReport.java]: code style

File: src/test/java/hudson/plugins/performance/build/PerformanceTestBuildTest.java [hunk 4]
```diff
@@ -283,10 +287,10 @@ public void testGenerateReportInPipe() throws Exception {
         String jobLog = new String(stream.toByteArray());
 
         File reportFile = new File(run.getRootDir(), "performance-reports/Taurus/aggregate-results.xml");
-        assertTrue(jobLog, jobLog.contains("Performance: Recording Taurus reports"));
-        assertTrue(jobLog, jobLog.contains("aggregate-results.xml'"));
-        assertTrue(jobLog, jobLog.contains("Performance: Parsing JMeter report file '" + reportFile.getAbsolutePath() + "'."));
-        assertTrue(jobLog, reportFile.exists());
+        assertTrue("Report file "+ reportFile.getAbsolutePath()+" expected to exist", reportFile.exists());
```
Comment (artem-fedorov) [inline src/test/java/hudson/plugins/performance/build/PerformanceTestBuildTest.java]: code style

File: src/main/java/hudson/plugins/performance/reports/PerformanceReport.java (hunk 2 above)
Comment (pmouawad) [reply src/main/java/hudson/plugins/performance/reports/PerformanceReport.java]: can you point me to codestyle rules  ?Thanks

Comment (artem-fedorov) [reply src/main/java/hudson/plugins/performance/reports/PerformanceReport.java]: you can use this rules https://wiki.apache.org/jmeter/CodeStyleGuidelines

**Comment by pmouawad** [comment]
Hello @artem-fedorov, @undera,
I continued improving PR by fixing codacy reported issues as long as other ones noted in comments of every commit.
I made tests of Relative and non relative constraints and it works fine.

I hope there is no remaining issue forbidding merge.

Regards 
Philippe

**Comment by artem-fedorov** [comment]
Awesome refactoring Philippe!  @pmouawad

File: src/main/java/hudson/plugins/performance/PerformancePublisher.java [hunk 5]
```diff
@@ -134,7 +129,7 @@
     /**
      * @deprecated as of 1.3. for compatibility
      */
-    private transient String filename;
+    private String filename;
```
[the remark this answers is not in the document]
Comment (pmouawad) [reply src/main/java/hudson/plugins/performance/PerformancePublisher.java]: Hello @artem-fedorov ,

Can you make the modification ?
I am not sure to be able to do it today.

Thanks

[the remark this answers is not in the document]
Comment (artem-fedorov) [reply src/main/java/hudson/plugins/performance/PerformancePublisher.java]: Don't worry, I'll wait you. Also PR has a merge conflicts now

**Comment by pmouawad** [comment]
Hi @artem-fedorov ,

Fixed.
As you can see the more we wait the more merge conflict can appear, so I found some time to take into account your notes.
Thanks
Regards

**Comment by artem-fedorov** [comment]
@pmouawad We had a release a few days ago. Do you have any plans for contribute new features soon?

**Comment by pmouawad** [comment]
Hello @artem-fedorov ,

Using the plugin, I have a doubt about performance and stability of the feature that allows you to see every Sample result.
On a big load test, I think it would be very slow and potentially crash Jenkins in OOM.
So that was a potential next feature but no immediate plans and wanted to have your opinion on how we could be fixing this.

Regards

**Comment by artem-fedorov** [comment]
@pmouawad Is it possible to install snapshot from master to your Jenkins server?

**Comment by pmouawad** [comment]
@artem-fedorov , sorry for late reply , I just missed this reply.
Could you either merge PR https://github.com/jenkinsci/performance-plugin/pull/187 or revert https://github.com/jenkinsci/performance-plugin/pull/182

I had tested without PR 182 and it was working fine.

Thanks

**Comment by pmouawad** [comment]
Hello @artem-fedorov ,

When do you plan to release 3.16 ?

Thank you

**Comment by richmarshall68** [comment]
I am also interested in knowing the date of release for 3.16
thx, Rich Marshall

**Comment by artem-fedorov** [comment]
hi @pmouawad  @richmarshall68 
Performance plugin 3.16 has been released

**Comment by pmouawad** [comment]
Большое спасибо!

**Pull request merged**